test/functional/api/buffer_spec.lua

```diff
@@ -577,6 +578,43 @@ describe('api/buf', function()
       ]])
       eq({'one', 'two'}, get_lines(0, 2, true))
     end)
+
+    it('updates folds #19571', function()
+      local screen = Screen.new(40, 5)
+      screen:attach()
+      local lines = {
+        '',
+        'FOLD START',
+        '',
+        'FOLD END',
+        '',
+      }
+      set_lines(0, -1, false, lines)
+      command('2,4fold')
+      screen:expect{grid=[[
+        ^                                        |
+        +--  3 lines: FOLD START················|
+                                                |
+        ~                                       |
+                                                |
+      ]]}
+      set_lines(4, 5, false, {})
+      screen:expect{grid=[[
+        ^                                        |
+        +--  3 lines: FOLD START················|
+        ~                                       |
+        ~                                       |
+                                                |
+      ]]}
+      set_lines(0, -1, false, lines)
+      screen:expect{grid=[[
+        ^                                        |
+        FOLD START                              |
+                                                |
+        FOLD END                                |
+                                                |
+      ]]}
+    end)
```

Comment:
This test does not use `nvim_buf_set_text`, which was used in the issue.

Reply:
The issue exists for both since they both call `mark_adjust` in the same way.